### ISSUE TITLE
Amplify drops event attributes, quota in AWS iOS SDK is different from PinPoint documentation

### ISSUE THREAD

**Issue by frankiesimon** [aws-amplify/amplify-swift#2471]
### Describe the bug

We started noticing that our events (sent from iOS client via Amplify to AWS PinPoint and forwarded with Kinesis to a lambda function) are missing random attribute fields.

After talking to both AWS support and `ruisebas` on the Amplify discord this seems to be the issue.

AWS iOS SDK drops attributes from an event if the combined number of metrics and attributes is 50 or above:
https://github.com/aws-amplify/aws-sdk-ios/blob/main/AWSPinpoint/AWSPinpointEvent.m#L114-L126

Meanwhile AWS PinPoint has a documented quota of 40 attributes/metrics which is apparently checked separately:
https://docs.aws.amazon.com/pinpoint/latest/developerguide/quotas.html#quotas-events

While the code removes "excess" attributes to fulfill "attributes+metrics<50", the actual payload can still fail if the number of *attributes* is *40* or above.

Here's debug output from such a case

(requires)
```
import AWSCore
...
AWSDDLog.sharedInstance.logLevel = .verbose
AWSDDLog.add(AWSDDTTYLogger.sharedInstance)
```

```
Response body:
{"Results":{"95B0C5F7-584B-44A7-B801-F77A6D736836":{"EndpointItemResponse":{"StatusCode":202,"Message":"Accepted"},"EventsItemResponse":{"525FD708-4105-4C9D-BF2E-7CF83AE386A6":{"StatusCode":400,"Message":"Number of attributes per event submitted should be less than 40."}}}}}

PutEventsResponse received: [<AWSPinpointTargetingPutEventsResponse: 0x2817c5c30> {
    eventsResponse = "<AWSPinpointTargetingEventsResponse: 0x2817c5ee0> {\n    results =     {\n        \"95B0C5F7-584B-44A7-B801-F77A6D736836\" = \"<AWSPinpointTargetingItemResponse: 0x2811b0ea0> {\\n    endpointItemResponse = \\\"<AWSPinpointTargetingEndpointItemResponse: 0x2811b0d60> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n    eventsItemResponse =     {\\n        \\\"525FD708-4105-4C9D-BF2E-7CF83AE386A6\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x2811b0ec0> {\\\\n    message = \\\\\\\"Number of attributes per event submitted should be less than 40.\\\\\\\";\\\\n    statusCode = 400;\\\\n}\\\";\\n    };\\n}\";\n    };\n}";
}]
EndpointProfile updated successfully.
Unable to successfully deliver event to server. Event will be saved with retry count += 1. Event id 525FD708-4105-4C9D-BF2E-7CF83AE386A6
Successfully put events to server--response code: 202. accepted: 0; retryable: 1; dirty: 0
```

It's unclear why the checks differ and also I find it extremely problematic that there is no way to know that the event I'm recording will have a corrupted payload with silent removal of fields.

In my production environment I saw events that do manage to get sent (they reach a lambda via a delivery stream) but with missing fields. I think that this is what happens:

* I record an event with over 50 attributes+metrics. For example: 20 metrics and 40 attributes
* Amplify removes 11 attributes so now there are 20 metrics and 29 attributes.
* GUESSTIMATE: Based on the error response body, AWSPinpoint looks at attributes separately, if there are less than 40 the event will come through with the dropped attributes.

### Steps To Reproduce

```swift
Steps to reproduce the behavior:

Use verbose level of AWSDDLog (code in issue description)

        
        var properties: AnalyticsProperties = [:]
        
        let numOfAttributes = 40
        if numOfAttributes > 0 {
            for i in 1...numOfAttributes {
                properties["my_\(i)_attribute"] = ""
            }
        }
        
        let numOfMetrics = 20
        if numOfMetrics > 0 {
            for i in 1...numOfMetrics {
                properties["my_\(i)_metric"] = 0
            }
        }
        let event = BasicAnalyticsEvent(name: "madeuptype", properties: properties)
        Amplify.Analytics.record(event: event)
```

Check debug output, random attributes will be dropped:
`Max number of attributes/metrics reached, dropping attribute with key: my_8_attribute`

Continue reading debug output, the payload will be successfully sent with 29 attributes and 20 metrics and pass the API.
```
(this is for a bunch of events, all with the same metrics/attributes as described above):

Response body:
{"Results":{"95B0C5F7-584B-44A7-B801-F77A6D736836":{"EndpointItemResponse":{"StatusCode":202,"Message":"Accepted"},"EventsItemResponse":{"6D7ED51F-AB43-49EE-8774-CBAE09B0E94F":{"StatusCode":202,"Message":"Accepted"},"809CD165-F1D2-4DA2-A4E6-5617C829B2AE":{"StatusCode":202,"Message":"Accepted"},"0DC1FE5A-7E06-4EE7-BCF3-13C59F7BF79D":{"StatusCode":202,"Message":"Accepted"},"DEB0B875-CE4A-4ED7-ABA8-13252988547B":{"StatusCode":202,"Message":"Accepted"},"817E270C-D8D8-47CC-A136-D4ED19F14C12":{"StatusCode":202,"Message":"Accepted"},"36B94363-7AD4-4340-B415-6A6D4E263877":{"StatusCode":202,"Message":"Accepted"},"8047E170-0D72-4227-BE8C-60AD79719EA8":{"StatusCode":202,"Message":"Accepted"},"54F0F110-CB10-4A4F-BCE7-CE2D02699B6F":{"StatusCode":202,"Message":"Accepted"},"52DA4392-F161-4E86-BC7F-73B5AA286F62":{"StatusCode":202,"Message":"Accepted"}}}}}
PutEventsResponse received: [<AWSPinpointTargetingPutEventsResponse: 0x281516310> {
    eventsResponse = "<AWSPinpointTargetingEventsResponse: 0x281516340> {\n    results =     {\n        \"95B0C5F7-584B-44A7-B801-F77A6D736836\" = \"<AWSPinpointTargetingItemResponse: 0x2813766e0> {\\n    endpointItemResponse = \\\"<AWSPinpointTargetingEndpointItemResponse: 0x281376220> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n    eventsItemResponse =     {\\n        \\\"0DC1FE5A-7E06-4EE7-BCF3-13C59F7BF79D\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x281374a60> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n        \\\"36B94363-7AD4-4340-B415-6A6D4E263877\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x281375be0> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n        \\\"52DA4392-F161-4E86-BC7F-73B5AA286F62\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x2813779a0> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n        \\\"54F0F110-CB10-4A4F-BCE7-CE2D02699B6F\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x281375a80> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n        \\\"6D7ED51F-AB43-49EE-8774-CBAE09B0E94F\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x281375b40> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n        \\\"8047E170-0D72-4227-BE8C-60AD79719EA8\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x281376860> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n        \\\"809CD165-F1D2-4DA2-A4E6-5617C829B2AE\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x281374740> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n        \\\"817E270C-D8D8-47CC-A136-D4ED19F14C12\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x281374640> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n        \\\"DEB0B875-CE4A-4ED7-ABA8-13252988547B\\\" = \\\"<AWSPinpointTargetingEventItemResponse: 0x281375b20> {\\\\n    message = Accepted;\\\\n    statusCode = 202;\\\\n}\\\";\\n    };\\n}\";\n    };\n}";
}]
```

### Expected behavior

Some clear way of knowing that fields are dropped. AND/OR a clear match between what is checked by PinPoint API and what is checked by the AWS iOS SDK code.

If the dropping of fields happens during the recording of the event, having a return value or callback or any way of knowing the event payload is modified would be helpful.

### Amplify Framework Version

1.28.1

### Amplify Categories

Analytics, Auth

### Dependency manager

Swift PM

### Swift version

5.6.1

### CLI version

10.0.0

### Xcode version

13.4.1 13F100

### Relevant log output

```shell
Described above.
```

### Is this a regression?

No

### Device

iPhone 8 Simulator / real device

### iOS Version

15.5

### Specific to simulators

No

**Comment by phantumcode** [comment]
Code fix has been merged in.  Fix will be available in the next release.

**Comment by frankiesimon** [comment]
@phantumcode Thank you for the swift response to this.

However, I am not sure I understand where the "combined number of metrics+attributes" quota is coming from. As far as I could tell (and maybe I need to debug again) - the attributes and metrics are checked separately. Is there any place where it's written that PinPoint cannot accept something like 30 attributes and 30 metrics in a single event?

Also - are the constants in Amplify used anywhere in the code or are they just there for public use to pre-check against these quotas?

**Comment by frankiesimon** [comment]
To reiterate - I do not see that PinPoint really has a quota for the **combined count** of attributes and metrics so Amplify could possibly prevent events from being sent even in cases PinPoint would accept them. It believe there's a separate quota for either attributes or metrics, each one can be 40 separately and unrelated to the other (again, requires testing which I can't do because Amplify would remove what it thinks is excess properties).

ALSO - while the AWS iOS SDK behavior for removing attributes/metrics is consistent because of the order of execution (first attributes, then metrics) - the behavior in Amplify will be undetermined since properties are stored in a dictionary and will be iterated in a random order, hence Amplify can remove a random selection of attributes and metrics. Meanwhile, AWS iOS SDK (PinPoint) code would have processed the attributes first (removing the ones that are over the quota), then metrics.

I was hoping to write code that mimics the filtering behavior to predict cases in which data will be lost (i.e. distinguish between event fields being dropped and the whole event being dropped) but because of the current implementation it's impossible to do so without some dangerous assumptions.

**Comment by ruisebas** [comment]
Hi @frankiesimon, thanks for the update.
I tried submitting events with 40 metrics and 40 attributes (for a total of 80 properties) and they were all accepted by Pinpoint, so I've asked them to clarify if this is the expected behaviour. Once I hear back I'll update the SDK PR and publish the Amplify ones to reflect the new values.

> Current code would drop the 40th metric/attribute at random. This should be corrected.

It would accept up to the 40th random property, and drop anything else. As you mentioned, this is an unfortunate byproduct of using a dictionary in `BasicAnalyticEvent` for properties. When they are iterated and added to the PinpointEvent's attributes and metrics (which happens on `record(event:)`), they are not sorted. 

We might explore adding in upcoming releases something like a new `PinpointAnalyticEvent` structure that can immediately report a failure to add new properties, but right now it's a known limitation.

**Comment by frankiesimon** [comment]
Hi again @ruisebas thanks for the clarification. Do you have an opinion about making the constants public so the numbers can be used by implementers to adjust their prepared events before calling record()?

**Comment by ruisebas** [comment]
Got confirmation from Pinpoint that its 40 for each, so 40 metrics and 40 attributes is fine 👍 .

>  Do you have an opinion about making the constants public 

I don't see any problem with that, I'll update the PRs to include them :)

**Comment by ruisebas** [comment]
This issue has been fixed in both v1 and v2:

- For v2.x, in release [**2.0.1**](https://github.com/aws-amplify/amplify-swift/releases/tag/v2.0.1)
- For v1.x, starting from release **1.28.1**. The fix was included in **AWS SDK for iOS** [2.28.1](https://github.com/aws-amplify/aws-sdk-ios/releases/tag/2.28.1), so just update your dependencies and validate that you get this version.
